### PR TITLE
PYIC-859: Add new SqsService to write messages to the audit event queue

### PR DESCRIPTION
**Pull request by dannyhayes-gds**
### What changed

- Create new common service to send messages to our SQSAuditEvent queue.
- Create a enum to hold each of the type events we will send.
- Send event after the CRI VC has been stored.

### Why did it change
For slice 2 we are sending an event to the TxMA SQS queue after a CRI VC has been gathered and validated.

### Issue tracking

- [PYIC-859](https://govukverify.atlassian.net/browse/PYIC-859)

### CODE REVIEW

File: lambdas/credentialissuerreturn/src/main/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandler.java [hunk 1]
```diff
@@ -66,11 +76,17 @@ public APIGatewayProxyResponseEvent handleRequest(
                             accessToken, credentialIssuerConfig, request.getIpvSessionId());
             credentialIssuerService.persistUserCredentials(verifiableCredential, request);
 
+            sqsService.sendAuditEvent(
+                    AuditEventTypes.IPV_CREDENTIAL_RECEIVED_AND_SIGNATURE_CHECKED);
+
```
Comment (Wynndow) [inline lambdas/credentialissuerreturn/src/main/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandler.java]: Just a thought, the event is "received and signature checked". So this should probably happen before the persistence. If something borks with writing it to Dynamdb we should still have created the audit event.

File: lambdas/credentialissuerreturn/src/test/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandlerTest.java [hunk 2]
```diff
@@ -72,7 +75,8 @@ void setUp() throws URISyntaxException {
     @Test
     void shouldReceive200AndJourneyResponseOnSuccessfulRequest() throws JsonProcessingException {
         CredentialIssuerReturnHandler handler =
-                new CredentialIssuerReturnHandler(credentialIssuerService, configurationService);
+                new CredentialIssuerReturnHandler(
+                        credentialIssuerService, configurationService, sqsService);
```
Comment (Wynndow) [inline lambdas/credentialissuerreturn/src/test/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandlerTest.java]: Might be good to assert in this test that the sqsService has been called correctly?

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import com.amazonaws.services.sqs.model.SendMessageRequest;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import uk.gov.di.ipv.core.library.auditing.AuditEvent;
+import uk.gov.di.ipv.core.library.auditing.AuditEventTypes;
+import uk.gov.di.ipv.core.library.exceptions.SqsException;
+
+import java.time.Instant;
+
+public class SqsService {
```
Comment (kerrr) [inline lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java]: I think this might be better being `AuditService` or something like that?

File: lambdas/credentialissuerreturn/src/main/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandler.java (hunk 1 above)
Comment (dannyhayes-gds) [reply lambdas/credentialissuerreturn/src/main/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandler.java]: Yeah I was a little unsure of wether it should be before or after the persistence. I went with after as I thought it could be hard to follow if we've got an audit trail saying that we "received" a VC but then there being no record of it in the DB if there were DynamoDb issues.

But I see your point as well so I'm happy to change this

File: lambdas/credentialissuerreturn/src/test/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandlerTest.java (hunk 2 above)
Comment (dannyhayes-gds) [reply lambdas/credentialissuerreturn/src/test/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandlerTest.java]: Yeah good shout 👍

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java (hunk 3 above)
Comment (dannyhayes-gds) [reply lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java]: Yeah true, I'll change it 👍

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+            sqs.sendMessage(sendMessageRequest);
+        } catch (JsonProcessingException e) {
+            throw new SqsException(e);
+        }
+    }
+
+    private String generateMessageBody(AuditEventTypes eventType) throws JsonProcessingException {
+        AuditEvent auditEvent =
+                new AuditEvent(Instant.now().getEpochSecond(), eventType.toString());
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java]: If you changed the `eventName` field in the `AuditEvent` class to type `AuditEventTypes`, could you pass the event type in directly? I think that the object mapper would serialize it to the name of the enum which is what we want here? Save calling `toString` here?

Comment (dannyhayes-gds) [reply lib/src/main/java/uk/gov/di/ipv/core/library/service/SqsService.java]: Yeah I think you're right. I'll give it a try, thanks

File: lambdas/credentialissuerreturn/src/main/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandler.java (hunk 1 above)
Comment (Wynndow) [reply lambdas/credentialissuerreturn/src/main/java/uk/gov/di/ipv/core/credentialissuerreturn/CredentialIssuerReturnHandler.java]: Yeah it's a funny one. I would hope that if something did happen with the DB though that we'd catch an error and log it, so we should be able to work out what happened.

File: lib/src/main/java/uk/gov/di/ipv/core/library/auditing/AuditEvent.java [hunk 5]
```diff
@@ -11,12 +11,12 @@ public class AuditEvent {
     @JsonProperty private long timestamp;
 
     @JsonProperty("event_name")
-    private String eventName;
+    private AuditEventTypes eventName;
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/core/library/auditing/AuditEvent.java]: Tiny thing, but maybe this should be called just `event` now?

**Review by Wynndow — APPROVED** [review]

**Review by kerrr — APPROVED** [review]
